targets/TARGET_STM/i2c_api.c

```diff
@@ -885,9 +1047,27 @@ void HAL_I2C_MasterRxCpltCallback(I2C_HandleTypeDef *hi2c)
     /* Get object ptr based on handler ptr */
     i2c_t *obj = get_i2c_obj(hi2c);
     struct i2c_s *obj_s = I2C_S(obj);
+    I2C_HandleTypeDef *handle = &(obj_s->handle);
 
     /* Set event flag */
     obj_s->event = I2C_EVENT_TRANSFER_COMPLETE;
+
+#if DEVICE_I2C_ASYNCH
+    /* */
+    if (obj->handler == NULL) {
+      return;
+    }
+
+    i2c_async_event_t event;
+    event.sent_bytes     = obj->tx_buff.length;
+    event.received_bytes = (obj->rx_buff.length - handle->XferCount);
+    event.error          = false;
+
+    obj->handler(obj, &event, obj->ctx);
+
+    obj->handler = NULL;
```

Comment:
Why is obj->handler set to NULL just after being called ?
will it be set again by the I2C driver or application ?

Reply:
The callback is set when the read/write transfers are initialised, it's cleared when the transfer is ended. Any new transfers will initialise a new callback.